### PR TITLE
GLFW support for C

### PR DESCRIPTION
**Pull request by grovesNL**
- Add surface creation from a CAMetalLayer on macOS (tested with macOS)
- Add surface creation from hinstance/hwnd on Windows (tested with Vulkan and DX12)
- Update hello_triangle_c to match hello_triangle_rust
- Change from Makefile to CMake for C example
- C example uses GLFW and expects it to be installed somewhere, linking is left to the user although it's not too difficult
- Update bindings

### CODE REVIEW

**Comment by grovesNL** [comment]
This Makefile is becoming a bit of a mess when trying to add the Windows target, I think I'm going to switch to cmake instead.

**Comment by kvark** [comment]
It would be unfortunate to depend on glfw the submodule (as well as CMake) for this project... Would it help if we make another repo for GLFW/SDL/other examples? They could technically be portable between wgpu-native and Dawn.

**Comment by grovesNL** [comment]
Sure, I have no preference about whether we should split examples into another repo. I guess it's nice to have the hello_world C and Rust examples side-by-side so that could be a downside. Maybe it wouldn't be too bad if we add a link to the examples in this repository's README or something.

For glfw I was thinking I could remove the submodule and only link to it if it's already installed, but on the other hand, it's not a great experience if people can't build the project because they don't have glfw already installed :) I was considering CMake because it seems relatively difficult to get cross-platform C builds (mostly Windows/Visual Studio) working consistently without a Makefile generator like CMake.

**Comment by grovesNL** [comment]
@kvark what are your thoughts on the updated PR? The CMake change is limited to replacing the Makefile with CMakeLists.txt, and GLFW isn't included by submodule anymore (assumes user can take care of the paths to link it on Windows, on macOS it's fairly easy)

File: wgpu-native/src/instance.rs [hunk 1]
```diff
@@ -67,6 +63,74 @@ pub extern "C" fn wgpu_instance_create_surface_from_winit(
         .register(surface)
 }
 
+#[allow(unused)]
+#[no_mangle]
+pub extern "C" fn wgpu_instance_create_surface_from_macos_layer(
```
Comment (kvark) [inline wgpu-native/src/instance.rs]: any reason this entry point isn't hidden for non-macos targets?

Comment (grovesNL) [reply wgpu-native/src/instance.rs]: To unconditionally include in wgpu.h for now, there are other ways we could work around it though

**Comment by kvark** [comment]
Merged in #49

**Pull request closed without merging**